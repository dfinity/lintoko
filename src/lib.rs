use std::{collections::HashSet, fs, io::Write, path::Path};

use anyhow::{Context, Result, anyhow};
use miette::{LabeledSpan, NamedSource, Report, Severity, miette};
use regex::Regex;
use serde::Deserialize;
use tracing::debug;
use tree_sitter::{Node, Parser, Query, QueryCapture, QueryCursor, Range, StreamingIterator};

#[derive(Debug, Deserialize)]
pub struct Rule {
    name: String,
    description: String,
    query: String,
}

pub fn default_rules() -> Vec<Rule> {
    vec![
        toml::from_str(include_str!("../default-rules/no-flexible.toml"))
            .expect("Failed to parse no-flexible rule"),
        toml::from_str(include_str!("../default-rules/no-stable.toml"))
            .expect("Failed to parse no-stable rule"),
        toml::from_str(include_str!("../default-rules/only-persistent-actor.toml"))
            .expect("Failed to parse only-persistent-actor rule"),
        toml::from_str(include_str!("../default-rules/pun-fields.toml"))
            .expect("Failed to parse pun-fields rule"),
        toml::from_str(include_str!("../default-rules/no-bool-switch.toml"))
            .expect("Failed to parse no-bool-switch rule"),
        toml::from_str(include_str!("../default-rules/assign-plus.toml"))
            .expect("Failed to parse assign-plus rule"),
        toml::from_str(include_str!("../default-rules/assign-minus.toml"))
            .expect("Failed to parse assign-minus rule"),
        toml::from_str(include_str!("../default-rules/unneeded-return.toml"))
            .expect("Failed to parse unneeded-return rule"),
    ]
}

pub fn load_rule_from_file(path: &Path) -> Result<Rule> {
    let content = std::fs::read_to_string(path)?;
    let rule = toml::from_str(&content)?;
    Ok(rule)
}

pub fn load_rules_from_directory(dir: &Path) -> Result<Vec<Rule>> {
    let mut rules = vec![];
    let entries = fs::read_dir(dir)
        .with_context(|| anyhow!("Failed to read rules from {}", dir.display()))?;
    for entry in entries {
        let entry = entry.with_context(|| anyhow!("Invalid entry"))?;
        let path = entry.path();
        if path.is_file() && path.extension().unwrap_or_default() == "toml" {
            debug!("Parsing extra rule at: {}", path.display());
            let rule = load_rule_from_file(&path)
                .with_context(|| anyhow!("Failed to parse rule from: '{}'", path.display()))?;
            rules.push(rule)
        }
    }
    Ok(rules)
}

/// Allows mentioning captures in rule descriptions, and templates them with the actual captures when reporting errors.
fn template_description(
    template: &str,
    query: &Query,
    captures: &[QueryCapture<'_>],
    input: &str,
) -> Result<String> {
    let re = Regex::new(r"@([a-z-]+)").unwrap();
    let mut new = String::with_capacity(template.len());
    let mut last_match = 0;
    for caps in re.captures_iter(template) {
        let m = caps.get(0).unwrap();
        let name = &caps[1];
        new.push_str(&template[last_match..m.start()]);
        let capture = captures.iter().find(|c| query.capture_names()[c.index as usize] == name).with_context(|| {
            anyhow!("Failed to find capture with name '{name}', when templating error description:\n\n'{template}'")
        })?;
        new.push_str(
            capture
                .node
                .utf8_text(input.as_bytes())
                .context("Non utf-8 text input")?,
        );
        last_match = m.end();
    }
    new.push_str(&template[last_match..]);
    Ok(new)
}

fn apply_rule(rule: &Rule, tree: Node, input: &str) -> Result<Vec<Report>> {
    let query = Query::new(&tree_sitter_motoko::LANGUAGE.into(), &rule.query)
        .with_context(|| format!("Failed to create query for rule '{}'", rule.name))?;
    let error_capture_index = query.capture_index_for_name("error").with_context(|| {
        anyhow!(
            "Expected query to contain `@error` captures:\n{}",
            rule.query
        )
    })?;
    let filter_capture_index = query.capture_index_for_name("filter");
    let trailing_capture_index = query.capture_index_for_name("trailing");

    let mut cursor = QueryCursor::new();
    let mut matches = cursor.matches(&query, tree, input.as_bytes());
    let mut filtered: HashSet<Range> = HashSet::new();
    let mut errors = Vec::new();
    while let Some(m) = matches.next() {
        // Works around a tree-sitter bug that doesn't let us use trailing anchors: https://github.com/tree-sitter/tree-sitter/issues/4558
        if let Some(trailing_capture_index) = trailing_capture_index
            && m.nodes_for_capture_index(trailing_capture_index)
                .any(|n| n.next_named_sibling().is_some())
        {
            continue;
        };
        for error_node in m.nodes_for_capture_index(error_capture_index) {
            // NOTE: We have to use `to_vec` here, or tree-sitter will silently swap the captures under our feet.
            errors.push((error_node.range(), m.captures.to_vec()));
        }

        if let Some(filter_capture_index) = filter_capture_index {
            for filter_node in m.nodes_for_capture_index(filter_capture_index) {
                filtered.insert(filter_node.range());
            }
        }
    }
    let mut seen = HashSet::new();
    let mut diagnostics = vec![];
    for (range, captures) in errors {
        if filtered.contains(&range) {
            continue;
        }
        // Avoid reporting the same diagnostic twice on the same range
        if !seen.insert(range) {
            continue;
        }
        let description = template_description(&rule.description, &query, &captures, input)?;
        let diagnostic = miette!(
            severity = Severity::Error,
            labels = vec![LabeledSpan::new_primary_with_span(
                Some(description),
                (range.start_byte, range.end_byte - range.start_byte)
            )],
            "[ERROR]: {}",
            rule.name
        );
        diagnostics.push(diagnostic);
    }
    Ok(diagnostics)
}

pub fn lint_file(path: &str, input: &str, rules: &[Rule], mut out: impl Write) -> Result<usize> {
    let mut parser = Parser::new();
    parser
        .set_language(&tree_sitter_motoko::LANGUAGE.into())
        .expect("Error loading Motoko grammar");
    let tree = parser.parse(input.as_bytes(), None).unwrap();
    let mut diagnostics = Vec::new();
    for rule in rules {
        diagnostics.extend(apply_rule(rule, tree.root_node(), input)?);
    }
    let count = diagnostics.len();
    diagnostics.sort_by_key(|d| d.labels().unwrap().next().unwrap().offset());
    for diagnostic in diagnostics {
        let pretty = diagnostic.with_source_code(NamedSource::new(path, input.to_string()));
        writeln!(&mut out, "{pretty:?}")?
    }
    Ok(count)
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn no_rules() {
        let mut out: Vec<u8> = vec![];
        let err_count = lint_file(
            "<input_path>",
            include_str!("../test-data.mo"),
            &[],
            &mut out,
        )
        .unwrap();
        assert_eq!(err_count, 0);
        assert_eq!(str::from_utf8(&out).unwrap(), "");
    }

    #[test]
    fn it_lints_default_rules() {
        let mut out: Vec<u8> = vec![];
        unsafe {
            std::env::set_var("NO_COLOR", "1");
        }
        let rules = default_rules();
        let _err_count = lint_file(
            "<input_path>",
            include_str!("../test-data.mo"),
            &rules,
            &mut out,
        )
        .unwrap();
        let lint_output = str::from_utf8(&out).unwrap();
        insta::assert_snapshot!(lint_output);
    }

    #[test]
    fn it_lints_custom_rules() {
        let mut out: Vec<u8> = vec![];
        unsafe {
            std::env::set_var("NO_COLOR", "1");
        }
        let mut rules = default_rules();
        rules.extend(load_rules_from_directory(Path::new("custom-rules")).unwrap());
        let _err_count = lint_file(
            "<input_path>",
            include_str!("../test-data.mo"),
            &rules,
            &mut out,
        )
        .unwrap();
        let lint_output = str::from_utf8(&out).unwrap();
        insta::assert_snapshot!(lint_output);
    }
}
