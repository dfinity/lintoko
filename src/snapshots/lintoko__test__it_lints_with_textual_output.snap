---
source: src/lib.rs
expression: lint_output
---
<input_path>:1:0 Error: Do not use the Result pattern. Use Runtime.trap() instead.
Found in:
1 import Result "mo:base/Result";

<input_path>:3:0 Error: Do not use data structures from pure/*. Use imperative structures instead.
Found in:
3 import PureList "mo:core/pure/List";

<input_path>:5:7 Error: Do not use `shared(msg)`. Use `shared({ caller })` instead.
Found in:
5 shared (msg) actor class() {

<input_path>:7:4 Error: Do not use let-else. Use a switch instead.
Found in:
7     let ?x = null else { return };

<input_path>:10:2 Error: Do not use the `flexible` keyword. Use `transient` instead.
Found in:
10   flexible let flexibleLet = 42;

<input_path>:11:2 Error: Do not use the `stable` keyword. It's implicit when using persistent actors.
Found in:
11   stable let stableLet = 42;

<input_path>:15:15 Error: Do not use `shared(msg)`. Use `shared({ caller })` instead.
Found in:
15   public shared(msg) func sharedCaller() {};

<input_path>:16:16 Error: Do not use `shared(msg)`. Use `shared({ caller })` instead.
Found in:
16   public shared msg func sharedCallerVar() {};

<input_path>:21:2 Error: Only use `transient` for OrderedMap.Make, OrderedSet.Make, and Random.Finite
Found in:
21   transient let transientMap = OrderedMap.Make.Well<Nat>(Nat.compare);

<input_path>:23:2 Error: Use `x += 1` instead of reassigning the variable.
Found in:
23   x := 1 + x;

<input_path>:24:2 Error: Use `x += 2` instead of reassigning the variable.
Found in:
24   x := x + 2;

<input_path>:26:2 Error: Use `x -= 1` instead of reassigning the variable
Found in:
26   x := x - 1;

<input_path>:29:2 Error: Use `x *= 1` instead of reassigning the variable.
Found in:
29   x := x * 1;

<input_path>:30:2 Error: Use `x *= 1` instead of reassigning the variable.
Found in:
30   x := 1 * x;

<input_path>:32:2 Error: Use `x /= 1` instead of reassigning the variable
Found in:
32   x := x / 1;

<input_path>:35:2 Error: Use `x #= "1"` instead of reassigning the variable
Found in:
35   x := x # "1";

<input_path>:38:12 Error: Use field punning to to avoid repetition: Replace `{ field = field }` with `{ field }`
Found in:
38   let _ = { field = field };

<input_path>:40:10 Error: Don't switch on boolean values, use if instead
Found in:
40   let _ = switch _ {
41     case (false) {};
42     case (true) {};
43   };

<input_path>:44:10 Error: Don't switch on boolean values, use if instead
Found in:
44   let _ = switch _ {
45     case true {};
46     case false {};
47   };

<input_path>:54:46 Error: Only return primitive types from publicly exposed endpoints
Found in:
54   public func listReturningFunction() : async List<Text> {

<input_path>:57:45 Error: Only return primitive types from publicly exposed endpoints
Found in:
57   public func setReturningFunction() : async Set.Set<Text> {

<input_path>:60:45 Error: Only return primitive types from publicly exposed endpoints
Found in:
60   public func mapReturningFunction() : async Map.Map<Text, Nat> {

<input_path>:73:4 Error: unneeded return statement. You can remove the `return`
Found in:
73     return 10

<input_path>:75:26 Error: unneeded return statement. You can remove the `return`
Found in:
75   func unneededReturn() = return 10;

<input_path>:79:6 Error: unneeded return statement. You can remove the `return`
Found in:
79       return 4;

<input_path>:86:14 Error: unneeded return statement. You can remove the `return`
Found in:
86     if (true) return 4

<input_path>:95:13 Error: unneeded return statement. You can remove the `return`
Found in:
95       case 1 return 40;

<input_path>:96:15 Error: unneeded return statement. You can remove the `return`
Found in:
96       case 2 { return 40; };

<input_path>:100:7 Error: Types should be capitalized and use CamelCase.
Found in:
100   type lowerCase = Nat;

<input_path>:101:7 Error: Types should be capitalized and use CamelCase.
Found in:
101   type Snake_case = Nat;

<input_path>:102:17 Error: Types should be capitalized and use CamelCase.
Found in:
102   type CamelCase<lowerCase, Snake_case> = Nat;

<input_path>:102:28 Error: Types should be capitalized and use CamelCase.
Found in:
102   type CamelCase<lowerCase, Snake_case> = Nat;

<input_path>:105:7 Error: Functions should start lower case and use camelCase.
Found in:
105   func UpperCase() {};

<input_path>:106:7 Error: Functions should start lower case and use camelCase.
Found in:
106   func snake_case() {};

<input_path>:108:8 Error: Types should be capitalized and use CamelCase.
Found in:
108   class lowerCase() {};

<input_path>:109:8 Error: Types should be capitalized and use CamelCase.
Found in:
109   class Snake_Cased_ish() {};

<input_path>:110:18 Error: Types should be capitalized and use CamelCase.
Found in:
110   class CamelCase<lowerCase, Snake_case>() {};

<input_path>:110:29 Error: Types should be capitalized and use CamelCase.
Found in:
110   class CamelCase<lowerCase, Snake_case>() {};
